mullvad-relay-selector/tests/relay_selector.rs

```diff
@@ -0,0 +1,1124 @@
+                        std::net::IpAddr::V4(_) => talpid_types::net::IpVersion::V4,
+                        std::net::IpAddr::V6(_) => talpid_types::net::IpVersion::V6,
+                    }));
+                assert!(query
+                    .wireguard_constraints
+                    .port
+                    .matches_eq(&endpoint.peer.endpoint.port()));
+                assert!(match query.wireguard_constraints.obfuscation {
+                    SelectedObfuscation::Auto => true,
+                    SelectedObfuscation::Off => obfuscator.is_none(),
+                    SelectedObfuscation::Udp2Tcp => obfuscator.is_some(),
+                });
+            }
+            GetRelay::OpenVpn {
+                endpoint, bridge, ..
+            } => {
+                if BridgeQuery::should_use_bridge(&query.openvpn_constraints.bridge_settings) {
+                    assert!(bridge.is_some(), "Relay selector should have selected a bridge for query {query:?}, but bridge was `None`");
+                };
+                assert!(query
+                    .openvpn_constraints
+                    .port
+                    .map(|transport_port| transport_port.port.matches_eq(&endpoint.address.port()))
+                    .unwrap_or(true),
+                    "The query {query:?} defined a port to use, but the chosen relay endpoint did not match that port number.
+                    Expected: {expected}
+                    Actual: {actual}",
+                    expected = query.openvpn_constraints.port.unwrap().port.unwrap(), actual = endpoint.address.port()
+                );
+
+                assert!(query.openvpn_constraints.port.map(|transport_port| transport_port.protocol == endpoint.protocol).unwrap_or(true),
+                    "The query {query:?} defined a transport protocol to use, but the chosen relay endpoint did not match that transport protocol.
+                    Expected: {expected}
+                    Actual: {actual}",
+                    expected = query.openvpn_constraints.port.unwrap().protocol, actual = endpoint.protocol
+                );
+            }
+            GetRelay::Custom(_) => unreachable!(),
+        }
+    }
+
+    let relay_selector = default_relay_selector();
+    let tunnel_protocols: Vec<Constraint<TunnelType>> = RETRY_ORDER
+        .iter()
+        .map(|relay| relay.tunnel_protocol)
+        .collect();
+    for (retry_attempt, tunnel_protocol) in tunnel_protocols.iter().enumerate() {
+        match tunnel_protocol {
+            Constraint::Any => continue,
+            Constraint::Only(expected_tunnel_protocol) => {
+                let relay = relay_selector.get_relay(retry_attempt).unwrap_or_else(|_| {
+                    panic!("Retry attempt {retry_attempt} did not yield any relay")
+                });
+                let tunnel_type = tunnel_type(&unwrap_relay(relay));
+                assert_eq!(tunnel_type, *expected_tunnel_protocol);
+            }
+        }
+    }
```

Comment:
Is this part still needed, don't you check it at line 269 already?

Reply:
Right, that was an oversight on my part :blush: Nice catch!